waver/_dock_widget.py

```diff
@@ -0,0 +1,43 @@
+from magicgui import magic_factory
+from napari_plugin_engine import napari_hook_implementation
+from .simulation import Simulation
+
+
+@magic_factory(call_button="run",
+              spacing={'min': 1e-6, 'max':1},
+              speed={'min': 1, 'max':1e5},
+              time_step={'min': 1e-10, 'max':1},
+)
+def simulation(spacing: float=1e-3,
+               speed: float=343, 
+               time_step: float=200e-9) -> 'napari.types.LayerDataTuple':
+    """Run a single simulation.
+    """
+    # Should be provided by magicgui, right now hard coded
+    size = (12.8e-3, 12.8e-3) # how to do tuple - could be 1D, 2D, 3D?
+    spacing=1e-5
+    speed=343
+    time_step=200e-9
+    period=5e-6
+    duration=60e-6
+
+    # Create simulation
+    sim = Simulation(size=size, spacing=spacing, speed=speed, time_step=time_step)
+
+    # Add source
+    sim.add_source(location=(s/2 for s in size), period=period, ncycles=1)
+
+    # Add detector grid
+    sim.add_detector()
+
+    # Run simulation
+    sim.run(duration=duration)
```

Comment:
Note inside here is a tqdm progress bar

Reply:
Hmmm this is going to be a tricky one. 
If we rip out the tqdm inheritance, then I think it will be nigh on impossible to "hook into" the `tqdm` bar being constructed inside `run`.

If we keep it, then one option that we discussed at one stage, first suggested [here ](https://github.com/napari/napari/pull/2580#issuecomment-825811160) by @Carreau was writing a patch for `tqdm` so that if we suspect some code is using a `tqdm` progress bar, we actually get our `progress` object. I never ended up trying this, but that may be an option you could try - at least before we finish with the refactor and stop inheriting from `tqdm`.

Even without the patching, as the `progress` infrastructure currently stands, you could (in theory...), just swap your call to `tqdm` [here](https://github.com/sofroniewn/waver/blob/38e22f7e7e4e963a7dc193cddaa4e688a3b79b01/waver/simulation/simulation.py#L168) to a call to  `progress(range(self.time.nsteps), disable=not progress, leave=leave)` and, if a `viewer` was active when that call was made, it **should** show you the progress bar. If a viewer wasn't active you'd see the standard `tqdm` behaviour. I'd actually be very curious if that worked.

Now, let's say we do refactor `progress` and stop inheriting from `tqdm`, the only option I see would be change the same spot to `progress(range(self.time.nsteps))`. You wouldn't get any of the terminal `tqdm` functionality at this stage, because as we discussed at the meeting today it's hard to decide what to do if a `progress` object is built with no active `viewer`.

If you wanted 0 changes to your `run` function, I think you could only get an indeterminate progress bar. 

Honestly seeing this use case is making me wonder whether we should keep `tqdm` inheritance... 😂  The tests have been relatively stable in the last few PRs, and we could just merge the small PR for closing the progress bar semi-gracefully on viewer close. Plus we could down the line do the patching that I mentioned above, if we wanted... I don't know. I wonder what @tlambert03 thinks as well.